src/utils/structs/checkpoints.cairo

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.7.0 (utils/structs/checkpoints.cairo)
+
+use integer::u32_sqrt;
+use openzeppelin::utils::math;
+use serde::Serde;
+use super::storage_array::StorageArrayTrait;
+use super::storage_array::StorageArray;
+use traits::Into;
+
+/// `Trace` struct, for checkpointing values as they change at different points in
+/// time, and later looking up past values by block number. See {Votes} as an example.
+#[derive(Copy, Drop, starknet::Store)]
+struct Trace {
+    checkpoints: StorageArray<Checkpoint>
+}
+
+/// Generic checkpoint representation.
+#[derive(Copy, Drop, Serde, starknet::Store)]
+struct Checkpoint {
+    key: u64,
+    // TODO: Check if is worth implementing a u220 type in corelib for saving gas by packing.
+    // Maybe is worth using u128 specifically for the fee token, if the token has just 6 decimals.
+    value: u256
```

Comment:
felt252 has some limitations regarding math operations. As an example we need division, and felt252 Div is not implemented because from design the division of felts is different than the regular one, with the purpose of staying in the finite field.

After some comments from Starkware, we think is better to implement a packing mechanism instead of changing the type of the value member, using StorePacking from corelib. This will save us from unnecessary writings to storage even if the types of Checkpoint members stay as they are.

Reply:
> felt252 has some limitations regarding math operations

Ah right, there's that

> After some comments from Starkware, we think is better to implement a packing mechanism instead of changing the type of the value member, using StorePacking from corelib. This will save us from unnecessary writings to storage even if the types of Checkpoint members stay as they are.

Sounds good. This seems like the better approach

Reply:
Already implemented.